mkosi: add UsrOnly= setting for generating images with only a /usr/ p…

poettering opened the pull request:
…artition

This adds a new boolean option that changes our image layout slightly:
instead of include the whole root fs in the image, we just pack up the
/usr/ subtree.

This has two usecases:

1. Truly stateless systems that come up pristine on every single boot,
   where /etc and /var are populated by tmpfiles/sysusers and related
   calls

2. Systems that are shipped without root fs, but where systemd-repart
   adds one in, with a locally generated encryption key and so on. (this
   doesn't work fully yet, because of some initrd issues, will look at
   this later.)

A companion PR on the systemd side is: https://github.com/systemd/systemd/pull/18958

Fixes: #634

behrmann commented:
Looks good on first, look. I expect the unit tests to fail, because the new option needs to be added to the tests.

DaanDeMeyer commented on mkosi/__init__.py:
```diff
@@ -4757,6 +4790,9 @@ def create_parser() -> ArgumentParserMkosi:
         "--tmp-size", help="Set size of /var/tmp partition (only gpt_ext4, gpt_xfs, gpt_squashfs)", metavar="BYTES"
     )
+    group.add_argument(
+        "--usr-only", action=BooleanAction, help="Generate only a /usr/ partition, but not root partition"
```
"Generate a /usr/ partition instead of a root partition" ?

DaanDeMeyer commented on mkosi/__init__.py:
```diff
@@ -3585,26 +3607,37 @@ def make_minimal_btrfs(args: CommandLineArguments, root: str, for_cache: bool) -
+        return None
+
+    label = "usr" if args.usr_only else "root"
+    patched_root = (root + "/usr") if args.usr_only else root
```
Let's use a f-string here: `f"{root}/usr}"`

DaanDeMeyer commented on mkosi/__init__.py:
```diff
@@ -3585,26 +3607,37 @@ def make_minimal_btrfs(args: CommandLineArguments, root: str, for_cache: bool) -
-            # The --shrink option was added in btrfs-tools 4.14.1, before that it was the default behaviour.
-            # If the above fails, let's see if things work if we drop it
-            command.remove("--shrink")
+        command = ["mkfs.btrfs", "-L", label, "-d", "single", "-m", "single", "--rootdir", root, f.name]
```
Nit: we tend to name these variables `cmdline`.

behrmann commented on mkosi/__init__.py:
```diff
@@ -3585,26 +3607,37 @@ def make_minimal_btrfs(args: CommandLineArguments, root: str, for_cache: bool) -
+        command = ["mkfs.btrfs", "-L", label, "-d", "single", "-m", "single", "--rootdir", root, f.name]
+
+        if args.minimal:
+            command.append("--shrink")
+            try:
+                run(command)
+            except subprocess.CalledProcessError:
+                # The --shrink option was added in btrfs-tools 4.14.1, before that it was the default behaviour.
+                # If the above fails, let's see if things work if we drop it
+                command.remove("--shrink")
+                run(command)
```
Instead of adding and then removing it again, this might be clearer
```suggestion
            try:
                run(command + ["--shrink"])
            except subprocess.CalledProcessError:
                # The --shrink option was added in btrfs-tools 4.14.1, before that it was the default behaviour.
                # If the above fails, let's see if things work without it
                run(command)
```

behrmann reviewed (commented):
Only tangentially related

DaanDeMeyer commented on mkosi/__init__.py:
```diff
@@ -3585,26 +3607,37 @@ def make_minimal_btrfs(args: CommandLineArguments, root: str, for_cache: bool) -
+        if args.minimal:
+            command.append("--shrink")
+            try:
+                run(command)
```
This seems like it would clutter mkosi's output when --shrink is not supported. Would it be OK if we just grepped the --help output of mkfs.btrfs for --shrink and use that to determine whether we use it or not?

behrmann replied:
I'm not sure that's worth it here, even on Debian stretch btrfs-tools 4.20 is available via backports. The only distro for which this might be a problem (I haven't checked) could be CentOS 7, so this should be a rather rare occurrence.

poettering replied:
I remember I added this back because it was missing on the CI. 

If the CI is updated we could certainly drop the compat here, but maybe we should do this independently of this, this --shrink stuff is pre-existing code after all

DaanDeMeyer commented on mkosi/__init__.py:
```diff
@@ -3512,13 +3535,15 @@ def make_tar(args: CommandLineArguments, root: str, do_run_build_script: bool, f
     if for_cache:
         return None
 
+    tar_root_dir = (root + "/usr") if args.usr_only else root
```
Ditto for using a f-string (see below)

DaanDeMeyer commented on mkosi/__init__.py:
```diff
@@ -1324,7 +1347,7 @@ def prepare_root(args: CommandLineArguments, dev: Optional[str], cached: bool) -
         return
 
-    with complete_step("Formatting root partition"):
+    with complete_step("Formatting {} partition".format("usr" if args.usr_only else "root")):
```
f-string here as well (see below)

DaanDeMeyer reviewed (commented):
LGTM, just a few nits.

If I understand this correctly, one way of using this is that on the initial OS install, systemd-repart would create the root partition and the /usr partition and initialize it from the partition generated by mkosi. How would updates to the /usr partition work? Would we "download" those to some temp location in the rootfs and then run systemd-repart again which would overwrite the existing /usr partition with the new one (or the B partition when we're using an A/B scheme)?

(I also need to get to writing some small framework so we can easily add more integration tests aside from the few existing ones. Once all the pieces of this larger work are in place, it'd be super nice to have some kind of test in mkosi (or systemd I guess) verifying it all works end-to-end)

poettering commented:
> If I understand this correctly, one way of using this is that on the initial OS install, systemd-repart would create the root partition and the /usr partition and initialize it from the partition generated by mkosi. How would updates to the /usr partition work? Would we "download" those to some temp location in the rootfs and then run systemd-repart again which would overwrite the existing /usr partition with the new one (or the B partition when we're using an A/B scheme)?

So the model I am primarily going for is like this:

1. with mkosi build a minimal GPT image that consists of an ESP + a squashfs /usr + a verity for that /usr. dd that onto the disk you want to use.
2. On first boot, in the initrd, add an empty second pair of partitions for /usr + verity for it; then add an encrypted root fs on the remaining space, and maybe other partitions such as home or swap (all enrolled to TPM if available) — i.e. run systemd-repart
3. on first update find an empty pair of /usr + verity for it, download the new partitions into that; then download efi unified kernel and drop that into ESP. reboot
4. at boot sd-boot will now automatically pick the kernel with the newer version string and it's embedded command line ensures it can only boot the root partition/verity that matches the same version, configured via hash
5. on next update we do the same as on the first update, but because now both sets of partitions  are filled we first empty the oldest (i.e. the one with the lowest version string), and also delete any matching kernel from the ESP. Then we do the exact same as on the first update.

In other words: the initial image ships a partition table like this:

1. ESP (includes unified kernel v0.1)
2. squashfs, immutable /usr fs (with OS v0.1)
3. verity for /usr (with the right hashes for the OS v0.1)

Then on first boot this is extended to:

1. ESP (includes unified kernel v0.1)
2. /usr fs (with OS v0.1)
3. verity for /usr (with the right hashes for the OS v0.1)
4. /usr fs nr. 2 (empty for now)
5. verity for /usr nr .2 (empty for now)
6. luksified root fs (with tpm and all bells and whistles)

Then on first update this becomes:

1. ESP (includes unified kernels v0.1 + v0.2)
2. /usr fs (with OS v0.1)
3. verity for /usr (with the right hashes for the OS v0.1)
4. /usr fs nr. 2 (with OS v0.2)
5. verity for /usr nr .2 (with the right hashes for OS v.0.2)
6. luksified root fs (with tpm and all bells and whistles)

On the second update this becomes:

1. ESP (includes unified kernels v0.2 + v0.3)
2. /usr fs (with OS v0.3)
3. verity for /usr (with the right hashes for the OS v0.3)
4. /usr fs nr. 2 (with OS v0.2)
5. verity for /usr nr .2 (with the right hashes for OS v.0.2)
6. luksified root fs (with tpm and all bells and whistles)

And so on.

On factory reset we simply reformat then root fs, leave the other 5 partitions the way they are.

And a live-installer could boot up exactly like that and use repart to set up another disk, in which case we set things up exactly like in the first step, and stream the data for that from the currently booted /usr and reboot.

This is the primary model I am focussing on, but this is supposed to be generic of course, there are other models I also care for a lot, like one where we ship a full immutable, versioned rootfs and then only generate a luksified/writable /var on first boot.

most of the components of what i describe above already exist in one form of another. It's mostly jut filling the gaps. my work on the updater component is WIP on this branch btw: https://github.com/poettering/systemd/commits/sysupdate – it's pretty complete already, but currently supports downloads only via plain http, not yet casync.

> (I also need to get to writing some small framework so we can easily add more integration tests aside from the few existing ones. Once all the pieces of this larger work are in place, it'd be super nice to have some kind of test in mkosi (or systemd I guess) verifying it all works end-to-end)

Oh, yeah, tests would be good!

behrmann commented on mkosi/__init__.py:
```diff
@@ -3585,26 +3607,37 @@ def make_minimal_btrfs(args: CommandLineArguments, root: str, for_cache: bool) -
-            command.remove("--shrink")
+        command = ["mkfs.btrfs", "-L", label, "-d", "single", "-m", "single", "--rootdir", root, f.name]
+
+        if args.minimal:
```
This was flagged by mypy
```suggestion
        if args.minimize:
```

DaanDeMeyer commented:
Thanks for the detailed explanation! I'm Looking forward to having this available (in v249 maybe?)

poettering commented:
Force pushed a new version. Everything should be addressed but the --shrink thing, see comment above

poettering commented:
Updated again, incorporated @behrmann's fixes (thanks so much!) and did some minor additional fixes on the documentation. Let's see if this passes now.

poettering commented:
> Thanks for the detailed explanation! I'm Looking forward to having this available (in v249 maybe?)

I am trying for v249, but let's see... I need to address #625 first (and maybe #626) to actually have something to update with.

poettering commented:
Yay! It passed!
